--- git_org/arch/arm/Kconfig	2018-05-07 15:44:57.214247880 +0800
+++ git/arch/arm/Kconfig	2018-05-07 15:47:51.554226313 +0800
@@ -153,6 +153,10 @@
 	bool "Support Facebook's OpenBMC Wedge100 TOR"
 	select CPU_ARM926EJS
 
+config TARGET_BFMAVERICKS
+	bool "Support Barefoot's OpenBMC MAVERICKS TOR"
+	select CPU_ARM926EJS
+
 config TARGET_FBMINIPACK
        bool "Support Facebook's OpenBMC Minipack"
 	select CPU_ARM1176
--- git_org/board/fb-obmc/Kconfig	2018-05-07 15:44:58.130247814 +0800
+++ git/board/fb-obmc/Kconfig	2018-05-07 17:27:47.337484194 +0800
@@ -50,6 +50,19 @@
   default "fbwedge100"
 endif
 
+if TARGET_BFMAVERICKS
+config SYS_CPU
+  default "arm926ejs"
+config SYS_BOARD
+  default "ast-g4"
+config SYS_VENDOR
+  default "aspeed"
+config SYS_SOC
+  default "aspeed"
+config SYS_CONFIG_NAME
+  default "bfmavericks"
+endif
+
 if TARGET_FBMINIPACK
 config SYS_CPU
   default "arm1176"
--- git_org/board/fb-obmc/MAINTAINERS	2018-05-07 15:44:58.130247814 +0800
+++ git/board/fb-obmc/MAINTAINERS	2018-05-07 17:36:32.881419148 +0800
@@ -6,12 +6,14 @@
 F:	include/configs/fbyosemite.h
 F:	include/configs/fblightning.h
 F:	include/configs/fbwedge100.h
+F:	include/configs/bfmavericks.h
 F:	include/configs/fbwedge.h
 F:	include/configs/fbminipack.h
 F:	include/configs/fbttn.h
 F:	configs/fbyosemite_defconfig
 F:	configs/fblightning_defconfig
 F:	configs/fbwedge100_defconfig
+F:	configs/bfmavericks_defconfig
 F:	configs/fbwedge_defconfig
 F:	configs/fbminipack_defconfig
 F:	configs/fbttn_defconfig
--- git_org/board/facebook/Kconfig	2018-05-07 15:44:58.130247814 +0800
+++ git/board/facebook/Kconfig	2018-05-07 17:52:12.037302907 +0800
@@ -50,6 +50,19 @@
   default "fbwedge100"
 endif
 
+if TARGET_BFMAVERICKS
+config SYS_CPU
+  default "arm926ejs"
+config SYS_BOARD
+  default "ast-g4"
+config SYS_VENDOR
+  default "aspeed"
+config SYS_SOC
+  default "aspeed"
+config SYS_CONFIG_NAME
+  default "bfmavericks"
+endif
+
 if TARGET_FBCMM
 config SYS_CPU
   default "arm1176"
--- git_org/board/facebook/MAINTAINERS	2018-05-07 15:44:58.130247814 +0800
+++ git/board/facebook/MAINTAINERS	2018-05-07 17:53:13.121295347 +0800
@@ -6,10 +6,12 @@
 F:	include/configs/fbyosemite.h
 F:	include/configs/fblightning.h
 F:	include/configs/fbwedge100.h
+F:	include/configs/bfmavericks.h
 F:	include/configs/fbwedge.h
 F:	include/configs/fbttn.h
 F:	configs/fbyosemite_defconfig
 F:	configs/fblightning_defconfig
 F:	configs/fbwedge100_defconfig
+F:	configs/bfmavericks_defconfig
 F:	configs/fbwedge_defconfig
 F:	configs/fbttn_defconfig
