Title: Fix formating issues causing compiler errors
Date: 2018-03-29
Author: Amithash Prasad <amithash@fb.com>
Bug-Debian: 
Index: tpm-tools/src/tpm_mgmt/tpm_present.c
===================================================================
--- tpm-tools.orig/src/tpm_mgmt/tpm_present.c
+++ tpm-tools/src/tpm_mgmt/tpm_present.c
@@ -349,13 +349,13 @@ int main(int argc, char **argv)
 		}
 	} while (flags[++i].name);
 
-      out_success:
+out_success:
 	logSuccess(argv[0]);
 	iRc = 0;
-      out_close:
+out_close:
 	contextClose(hContext);
-      out:
-    if (szTpmPasswd && !isWellKnown)
-	shredPasswd( szTpmPasswd );
+out:
+	if (szTpmPasswd && !isWellKnown)
+		shredPasswd( szTpmPasswd );
 	return iRc;
 }
