From 1687057c5b63099a4ac391486841e7cf4a5d2c2a Mon Sep 17 00:00:00 2001
From: Igor Kanyuka <ifelmail@gmail.com>
Date: Mon, 2 Feb 2026 07:47:26 -0800
Subject: [PATCH] Subject: [PATCH] configure: lookup meson exutable from PATH

Upstream-Status: Inappropriate [workaround, would need a real fix for upstream]
---
 configure | 11 +----------
 1 file changed, 1 insertion(+), 10 deletions(-)

diff --git a/configure b/configure
index d8bc10060e..a91906a864 100755
--- a/configure
+++ b/configure
@@ -953,12 +953,6 @@ mkvenv="$python ${source_path}/python/scripts/mkvenv.py"
 $mkvenv ensuregroup --dir "${source_path}/python/wheels" \
      ${source_path}/pythondeps.toml meson || exit 1
 
-# At this point, we expect Meson to be installed and available.
-# We expect mkvenv or pip to have created pyvenv/bin/meson for us.
-# We ignore PATH completely here: we want to use the venv's Meson
-# *exclusively*.
-
-# for msys2
 get_pwd() {
     if pwd -W >/dev/null 2>&1; then
         pwd -W
@@ -967,10 +961,7 @@ get_pwd() {
     fi
 }
 
-meson="$(cd pyvenv/bin; get_pwd)/meson"
-if [ -f "$meson$EXESUF" ]; then
-  meson="$meson$EXESUF"
-fi
+meson=`which meson`
 
 # On Haiku, meson might show up in a "non-packaged" subfolder instead, see
 # https://github.com/haiku/haiku/blob/r1beta5/docs/user/storage/storageintro.dox
-- 
2.53.0-Meta

