From 91241535a53f8653a338d4c7f6f5d57bcd0af5c7 Mon Sep 17 00:00:00 2001
From: OpenBMC <openbmc@meta.com>
Date: Tue, 31 Mar 2026 06:55:47 -0700
Subject: [PATCH 10/10] hw/arm/aspeed: Add BMC machine to build

---
 hw/arm/meson.build | 1 +
 1 file changed, 1 insertion(+)

diff --git a/hw/arm/meson.build b/hw/arm/meson.build
index b187b946f0..3050a48200 100644
--- a/hw/arm/meson.build
+++ b/hw/arm/meson.build
@@ -56,6 +56,7 @@ arm_ss.add(when: 'CONFIG_ASPEED_SOC', if_true: files(
   'aspeed_ast2500_witherspoon.c',
   'aspeed_ast2500_yosemitev2.c',
   'aspeed_ast2600.c',
+  'aspeed_ast2600_anacapa.c',
   'aspeed_ast2600_bletchley.c',
   'aspeed_ast2600_catalina.c',
   'aspeed_ast2600_evb.c',
-- 
2.52.0

