From e45512fd4e843c40d5ac02996759fe7290c8f218 Mon Sep 17 00:00:00 2001
From: Peter Delevoryas <peter@pjd.dev>
Date: Fri, 15 Jul 2022 14:34:49 -0700
Subject: [PATCH] aspeed: Add grandteton-bmc

Signed-off-by: Peter Delevoryas <peter@pjd.dev>
Signed-off-by: Peter Delevoryas <pdel@fb.com>
---
 hw/arm/aspeed_ast2600_evb.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/hw/arm/aspeed_ast2600_evb.c b/hw/arm/aspeed_ast2600_evb.c
index a93275e7f1..ec448e6bc1 100644
--- a/hw/arm/aspeed_ast2600_evb.c
+++ b/hw/arm/aspeed_ast2600_evb.c
@@ -87,6 +87,17 @@ static const struct FBMachineData fb_machines[] = {
         .stdout_path = ASPEED_DEV_UART5,
         .ram_size    = FUJI_BMC_RAM_SIZE,
     },
+    {
+        .name        = "grandteton-bmc",
+        .desc        = "Facebook GrandTeton BMC (Cortex-A7)",
+        .soc_name    = "ast2600-a3",
+        .hw_strap1   = AST2600_EVB_HW_STRAP1,
+        .hw_strap2   = AST2600_EVB_HW_STRAP2,
+        .flash_model = "n25q00",
+        .macs_mask   = ASPEED_MAC3_ON,
+        .stdout_path = ASPEED_DEV_UART5,
+        .ram_size    = FUJI_BMC_RAM_SIZE,
+    },
     {
         .name        = "elbert-bmc",
         .desc        = "Facebook Elbert BMC (Cortex-A7)",
-- 
2.53.0-Meta

