From 1eeeabc04902d9b63d84c41c17f796c074fdb998 Mon Sep 17 00:00:00 2001
From: Tao Ren <taoren@meta.com>
Date: Mon, 10 Jul 2023 16:47:00 +0700
Subject: [PATCH] aspeed: add janga bmc

Signed-off-by: Tao Ren <taoren@meta.com>
---
 hw/arm/aspeed_ast2600_evb.c | 73 +++++++++++++++++++++++++++++++++++++
 1 file changed, 73 insertions(+)

diff --git a/hw/arm/aspeed_ast2600_evb.c b/hw/arm/aspeed_ast2600_evb.c
index adbcc31556..31430f6861 100644
--- a/hw/arm/aspeed_ast2600_evb.c
+++ b/hw/arm/aspeed_ast2600_evb.c
@@ -23,6 +23,10 @@
 
 #define FUJI_BMC_RAM_SIZE ASPEED_RAM_SIZE(2 * GiB)
 
+/* Janga hardware value */
+#define JANGA_BMC_HW_STRAP1    0x00000000
+#define JANGA_BMC_HW_STRAP2    0x00000000
+
 /* Montblanc hardware value */
 #define MONTBLANC_BMC_HW_STRAP1    0x00000000
 #define MONTBLANC_BMC_HW_STRAP2    0x00000000
@@ -40,6 +44,48 @@ static void ast2600_evb_i2c_init(AspeedMachineState *bmc)
                      TYPE_TMP105, 0x4d);
 }
 
+static const uint8_t janga_smb_fruid[] = {
+    0xFB, 0xFB, 0x04, 0xFF, 0x01, 0x05, 0x4A, 0x41, 0x4E, 0x47, 0x41, 0x02,
+    0x09, 0x32, 0x30, 0x2D, 0x31, 0x30, 0x30, 0x30, 0x31, 0x36, 0x03, 0x08,
+    0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x04, 0x0C, 0x31, 0x33,
+    0x32, 0x31, 0x30, 0x30, 0x30, 0x30, 0x35, 0x20, 0x20, 0x20, 0x05, 0x0C,
+    0x31, 0x33, 0x31, 0x31, 0x30, 0x30, 0x30, 0x30, 0x34, 0x20, 0x20, 0x20,
+    0x06, 0x0C, 0x52, 0x34, 0x30, 0x35, 0x34, 0x47, 0x30, 0x30, 0x30, 0x32,
+    0x30, 0x31, 0x07, 0x0D, 0x42, 0x30, 0x39, 0x31, 0x37, 0x32, 0x30, 0x34,
+    0x35, 0x30, 0x30, 0x32, 0x33, 0x08, 0x01, 0x01, 0x09, 0x01, 0x01, 0x0A,
+    0x01, 0x01, 0x0B, 0x08, 0x44, 0x55, 0x4D, 0x4D, 0x59, 0x20, 0x53, 0x4E,
+    0x0C, 0x03, 0x43, 0x4C, 0x53, 0x0D, 0x08, 0x32, 0x30, 0x32, 0x33, 0x30,
+    0x38, 0x32, 0x38, 0x0E, 0x03, 0x57, 0x55, 0x53, 0x0F, 0x03, 0x43, 0x54,
+    0x48, 0x10, 0x06, 0x01, 0x12, 0xFF, 0x91, 0xDB, 0xB4, 0x11, 0x06, 0x03,
+    0x12, 0xFF, 0x91, 0xDB, 0xB4, 0x12, 0x02, 0x65, 0x00, 0x13, 0x03, 0x53,
+    0x4D, 0x42, 0xFA, 0x02, 0x93, 0xBC,
+};
+
+static void janga_bmc_i2c_init(AspeedMachineState *bmc)
+{
+    AspeedSoCState *soc = bmc->soc;
+    I2CBus *i2c[16] = {};
+
+    for (int i = 0; i < 16; i++) {
+        i2c[i] = aspeed_i2c_get_bus(&soc->i2c, i);
+    }
+
+    at24c_eeprom_init_rom(i2c[3], 0x56, 8 * KiB, janga_smb_fruid, sizeof(janga_smb_fruid)); /* SMB_2 EEPROM */
+
+    /* CPLD and FPGA */
+    at24c_eeprom_init(i2c[1], 0x35, 256);  /* SMB_CPLD_1 aka SCM_CPLD */
+    at24c_eeprom_init(i2c[4], 0x35, 256);  /* COMe CPLD TODO: need to update */
+    at24c_eeprom_init(i2c[6], 0x35, 256); /* FCB#1 CPLD TODO: need to update*/
+    at24c_eeprom_init(i2c[2], 0x35, 256); /* FCB#2 CPLD TODO: need to update*/
+    at24c_eeprom_init(i2c[12], 0x60, 256); /* PWR_CPLD */
+    at24c_eeprom_init(i2c[13], 0x35, 256); /* IOB FPGA */
+
+    /* COMe Sensor/EEPROM */
+    at24c_eeprom_init(i2c[0], 0x56, 16 * KiB);          /* FRU EEPROM */
+    i2c_slave_create_simple(i2c[0], TYPE_LM75, 0x48); /* INLET Sensor */
+    i2c_slave_create_simple(i2c[0], TYPE_LM75, 0x4A); /* OUTLET Sensor */
+}
+
 static const uint8_t montblanc_scm_fruid[8192] = {
     0xfb, 0xfb, 0x04, 0xff, 0x01, 0x0d, 0x4d, 0x49, 0x4e, 0x49, 0x50, 0x41,
     0x43, 0x4b, 0x33, 0x5f, 0x53, 0x43, 0x4d, 0x02, 0x08, 0x32, 0x30, 0x30,
@@ -140,6 +186,27 @@ static void aspeed_machine_ast2600_evb_class_init(ObjectClass *oc,
     aspeed_machine_ast2600_class_emmc_init(oc);
 };
 
+#define JANGA_BMC_RAM_SIZE ASPEED_RAM_SIZE(2 * GiB)
+
+static void aspeed_machine_janga_class_init(ObjectClass *oc, const void *data)
+{
+    MachineClass *mc = MACHINE_CLASS(oc);
+    AspeedMachineClass *amc = ASPEED_MACHINE_CLASS(oc);
+
+    mc->desc = "Facebook Janga BMC (Cortex-A7)";
+    amc->soc_name = "ast2600-a3";
+    amc->hw_strap1 = JANGA_BMC_HW_STRAP1;
+    amc->hw_strap2 = JANGA_BMC_HW_STRAP2;
+    amc->fmc_model = "mx66l1g45g";
+    amc->spi_model = "mx66l1g45g";
+    amc->num_cs = 2;
+    amc->macs_mask = ASPEED_MAC3_ON;
+    amc->i2c_init = janga_bmc_i2c_init;
+    amc->uart_default = ASPEED_DEV_UART1;
+    mc->default_ram_size = JANGA_BMC_RAM_SIZE;
+    aspeed_machine_class_init_cpus_defaults(mc);
+};
+
 #define MONTBLANC_BMC_RAM_SIZE ASPEED_RAM_SIZE(2 * GiB)
 
 static void aspeed_machine_montblanc_class_init(ObjectClass *oc, const void *data)
@@ -168,6 +235,12 @@ static const TypeInfo aspeed_ast2600_evb_types[] = {
         .class_init    = aspeed_machine_ast2600_evb_class_init,
         .interfaces    = arm_machine_interfaces,
     },
+    {
+        .name          = MACHINE_TYPE_NAME("janga-bmc"),
+        .parent        = TYPE_ASPEED_MACHINE,
+        .class_init    = aspeed_machine_janga_class_init,
+        .interfaces    = arm_machine_interfaces,
+    },
     {
         .name          = MACHINE_TYPE_NAME("montblanc-bmc"),
         .parent        = TYPE_ASPEED_MACHINE,
-- 
2.53.0-Meta

