From 45fcf1724ab5e4fe6452dd8c35ad39d1130de067 Mon Sep 17 00:00:00 2001
From: Eli Huang <Eli_Huang@wiwynn.com>
Date: Fri, 5 Aug 2022 09:59:19 +0800
Subject: [PATCH 1/4] Revise aspeed 2600 dtsi

---
 arch/arm/boot/dts/aspeed-g6.dtsi | 42 ++++++++++++++++++++++++++++++++
 1 file changed, 42 insertions(+)

diff --git a/arch/arm/boot/dts/aspeed-g6.dtsi b/arch/arm/boot/dts/aspeed-g6.dtsi
index 1387a763a6a5..9845782b0157 100644
--- a/arch/arm/boot/dts/aspeed-g6.dtsi
+++ b/arch/arm/boot/dts/aspeed-g6.dtsi
@@ -323,6 +323,28 @@ apb {
 			#size-cells = <1>;
 			ranges;
 
+                        pwm_tach: pwm_tach@1e610000 {
+                                compatible = "aspeed,ast2600-pwm-tach", "simple-mfd", "syscon";
+                                reg = <0x1e610000 0x100>;
+                                clocks = <&syscon ASPEED_CLK_AHB>;
+                                resets = <&syscon ASPEED_RESET_PWM>;
+
+                                pwm: pwm {
+                                        compatible = "aspeed,ast2600-pwm";
+                                        #pwm-cells = <3>;
+                                        #address-cells = <1>;
+                                        #size-cells = <0>;
+                                        status = "disabled";
+                                };
+
+                                tach: tach {
+                                        compatible = "aspeed,ast2600-tach";
+                                        #address-cells = <1>;
+                                        #size-cells = <0>;
+                                        status = "disabled";
+                                };
+                        };
+
 			syscon: syscon@1e6e2000 {
 				compatible = "aspeed,ast2600-scu", "syscon", "simple-mfd";
 				reg = <0x1e6e2000 0x1000>;
@@ -363,6 +385,26 @@ scu_ic1: interrupt-controller@570 {
 				};
 			};
 
+                        jtag0: jtag@1e6e4000 {
+                                compatible = "aspeed,ast2600-jtag";
+                                reg = <0x1e6e4000 0x40 0x1e6e2040 0x8 0x1e6e2418 0x4>;
+                                clocks = <&syscon ASPEED_CLK_APB1>;
+                                resets = <&syscon ASPEED_RESET_JTAG_MASTER>;
+                                interrupts = <27>;
+                                status = "disabled";
+                        };
+
+                        jtag1: jtag@1e6e4100 {
+                                compatible = "aspeed,ast2600-jtag";
+                                reg = <0x1e6e4100 0x40 0x1e6e2050 0x8 0x1e6e2418 0x4>;
+                                clocks = <&syscon ASPEED_CLK_AHB>;
+                                resets = <&syscon ASPEED_RESET_JTAG_MASTER2>;
+                                interrupts = <GIC_SPI 53 IRQ_TYPE_LEVEL_HIGH>;
+                                pinctrl-names = "default";
+                                pinctrl-0 = <&pinctrl_jtagm_default>;
+                                status = "disabled";
+                        };
+
 			rng: hwrng@1e6e2524 {
 				compatible = "timeriomem_rng";
 				reg = <0x1e6e2524 0x4>;
-- 
2.35.1

