From cc31668f14fe0f6d1d80b276c1edb0425c7db94a Mon Sep 17 00:00:00 2001
From: WayneLiu-wiwynn <Wayne_SC_Liu@wiwynn.com>
Date: Mon, 13 Feb 2023 09:38:16 +0800
Subject: [PATCH] ARM dts aspeed greatlakes Add Facebookgreatlakes`

---
 arch/arm/boot/dts/Makefile | 1 +
 1 file changed, 1 insertion(+)

diff --git a/arch/arm/boot/dts/Makefile b/arch/arm/boot/dts/Makefile
index 2ee9c043218b..82641a81ce94 100644
--- a/arch/arm/boot/dts/Makefile
+++ b/arch/arm/boot/dts/Makefile
@@ -1589,6 +1589,7 @@ dtb-$(CONFIG_ARCH_ASPEED) += \
 	aspeed-bmc-facebook-elbert.dtb \
 	aspeed-bmc-facebook-fuji.dtb \
 	aspeed-bmc-facebook-galaxy100.dtb \
+	aspeed-bmc-facebook-greatlakes.dtb \
 	aspeed-bmc-facebook-minipack.dtb \
 	aspeed-bmc-facebook-tiogapass.dtb \
 	aspeed-bmc-facebook-wedge40.dtb \
-- 
2.25.1

