From 6fbbf1021b0da39d4f105d387007654e496e40a2 Mon Sep 17 00:00:00 2001
From: Tao Ren <taoren@meta.com>
Date: Fri, 24 Jul 2026 13:52:11 +0800
Subject: [PATCH] ARM dts aspeed fblite r1 add AMD APML device to i2c14

---
 .../aspeed/aspeed-bmc-facebook-fblite-r1.dts  | 22 +++++++++++++++++++
 1 file changed, 22 insertions(+)

diff --git a/arch/arm/boot/dts/aspeed/aspeed-bmc-facebook-fblite-r1.dts b/arch/arm/boot/dts/aspeed/aspeed-bmc-facebook-fblite-r1.dts
index b048dd5be..0a3c0ae5c 100644
--- a/arch/arm/boot/dts/aspeed/aspeed-bmc-facebook-fblite-r1.dts
+++ b/arch/arm/boot/dts/aspeed/aspeed-bmc-facebook-fblite-r1.dts
@@ -17,6 +17,14 @@ spi_gpio: spi {
 		num-chipselects = <1>;
 		cs-gpios = <&gpio0 ASPEED_GPIO(X, 0) GPIO_ACTIVE_LOW>;
 	};
+
+	/* Alert_L associated with socket 0 */
+	alertl_sock0 {
+		compatible = "apml-alertl";
+		status = "okay";
+		gpios = <&gpio0 ASPEED_GPIO(M, 3) GPIO_ACTIVE_LOW>;
+		socket-num = /bits/ 8 <0>;
+	};
 };
 
 &fmc {
@@ -39,6 +47,20 @@ &lpc_ctrl {
 	flash = <&spi1>;
 };
 
+&i2c14 {
+	status = "okay";
+	multi-master; 
+	sbtsi@4c {
+		compatible = "amd,sbtsi";
+		reg = <0x4c>;
+	};
+
+	sbrmi@3c {
+		compatible = "amd,sbrmi";
+		reg = <0x3c>;
+	};
+};
+
 &kcs1 {
 	status = "okay";
 	aspeed,lpc-io-reg = <0xca0>;
-- 
2.43.0

