diff --git a/meta-aspeed/recipes-bsp/u-boot/files/u-boot-v2013.07/include/configs/fbyosemite.h b/meta-aspeed/recipes-bsp/u-boot/files/u-boot-v2013.07/include/configs/fbyosemite.h
index 73a3cf9..397ca48 100644
--- a/meta-aspeed/recipes-bsp/u-boot/files/u-boot-v2013.07/include/configs/fbyosemite.h
+++ b/meta-aspeed/recipes-bsp/u-boot/files/u-boot-v2013.07/include/configs/fbyosemite.h
@@ -155,7 +155,7 @@
  */
 #define CONFIG_NR_DRAM_BANKS	1	   	/* we have 1 bank of DRAM */
 #define PHYS_SDRAM_1		  0x40000000 	/* SDRAM Bank #1 */
-#define PHYS_SDRAM_1_SIZE	0x8000000 	/* 128 MB */
+#define PHYS_SDRAM_1_SIZE	0x10000000 	/* 256 MB */
 
 #define CONFIG_SYS_SDRAM_BASE	0x40000000
 
