From 7a8cab044baaa8d003885e188d49e388d21b4caf Mon Sep 17 00:00:00 2001
From: peteryin <peter.yin@quantatw.com>
Date: Thu, 27 Jul 2023 10:02:41 +0800
Subject: [PATCH 1/2] fix-fan-enable-attribute

---
 lib/sysfs.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/lib/sysfs.c b/lib/sysfs.c
index 46bdcfc..a27adfd 100644
--- a/lib/sysfs.c
+++ b/lib/sysfs.c
@@ -371,6 +371,9 @@ sensors_subfeature_type sensors_subfeature_get_type(const char *name, int *nr)
 		if ((count = sscanf(name, matches[i].name, nr, &c))) {
 			/* Needed for matching pwm[1-*] */
 			if (matches[i].submatches == pwm_matches) {
+				if (c == '_') {
+					*nr=*nr+128;
+				}
 				return SENSORS_SUBFEATURE_PWM_IO;
 			}
 			break;
-- 
2.25.1

