From aeeb06de3c4cf578b019e93e4edbb7ff398331c0 Mon Sep 17 00:00:00 2001
From: peteryin <peter.yin@quantatw.com>
Date: Thu, 27 Jul 2023 10:03:21 +0800
Subject: [PATCH 2/2] modfiy-Sensors-precision-values

---
 prog/sensors/chips.c | 14 +++++++-------
 1 file changed, 7 insertions(+), 7 deletions(-)

diff --git a/prog/sensors/chips.c b/prog/sensors/chips.c
index f08351d..e447790 100644
--- a/prog/sensors/chips.c
+++ b/prog/sensors/chips.c
@@ -400,7 +400,7 @@ static void print_chip_in(const sensors_chip_name *name,
 	sf = sensors_get_subfeature(name, feature,
 				    SENSORS_SUBFEATURE_IN_INPUT);
 	if (sf && get_input_value(name, sf, &val) == 0)
-		printf("%+6.2f V  ", val);
+		printf("%+6.3f V  ", val);
 	else
 		printf("     N/A  ");
 
@@ -409,7 +409,7 @@ static void print_chip_in(const sensors_chip_name *name,
 			      sensors, &sensor_count, alarms, &alarm_count);
 
 	print_limits(sensors, sensor_count, alarms, alarm_count, label_size,
-		     "%s = %+6.2f V");
+		     "%s = %+6.3f V");
 
 	printf("\n");
 }
@@ -592,7 +592,7 @@ static void print_chip_power(const sensors_chip_name *name,
 
 	if (sf && get_input_value(name, sf, &val) == 0) {
 		scale_value(&val, &unit);
-		printf("%6.2f %sW%*s", val, unit, 2 - (int)strlen(unit), "");
+		printf("%6.3f %sW%*s", val, unit, 2 - (int)strlen(unit), "");
 	} else
 		printf("     N/A  ");
 
@@ -614,7 +614,7 @@ static void print_chip_power(const sensors_chip_name *name,
 		}
 	}
 	print_limits(sensors, sensor_count, alarms, alarm_count,
-		     label_size, "%s = %6.2f %s");
+		     label_size, "%s = %6.3f %s");
 
 	printf("\n");
 }
@@ -640,7 +640,7 @@ static void print_chip_energy(const sensors_chip_name *name,
 				    SENSORS_SUBFEATURE_ENERGY_INPUT);
 	if (sf && get_input_value(name, sf, &val) == 0) {
 		scale_value(&val, &unit);
-		printf("%6.2f %sJ", val, unit);
+		printf("%6.3f %sJ", val, unit);
 	} else
 		printf("     N/A");
 
@@ -803,7 +803,7 @@ static void print_chip_curr(const sensors_chip_name *name,
 	sf = sensors_get_subfeature(name, feature,
 				    SENSORS_SUBFEATURE_CURR_INPUT);
 	if (sf && get_input_value(name, sf, &val) == 0)
-		printf("%+6.2f A  ", val);
+		printf("%+6.3f A  ", val);
 	else
 		printf("     N/A  ");
 
@@ -812,7 +812,7 @@ static void print_chip_curr(const sensors_chip_name *name,
 			      sensors, &sensor_count, alarms, &alarm_count);
 
 	print_limits(sensors, sensor_count, alarms, alarm_count, label_size,
-		     "%s = %+6.2f A");
+		     "%s = %+6.3f A");
 
 	printf("\n");
 }
-- 
2.25.1

