From 91f30875c0b44a1e7eb3deecddcb2f0686d6021b Mon Sep 17 00:00:00 2001
From: Peter Delevoryas <pdel@fb.com>
Date: Mon, 23 May 2022 10:23:53 -0700
Subject: [PATCH 08/65] tests/avocado: Disable raspi2 initrd

Signed-off-by: Peter Delevoryas <pdel@fb.com>
---
 tests/avocado/boot_linux_console.py | 1 +
 1 file changed, 1 insertion(+)

diff --git a/tests/avocado/boot_linux_console.py b/tests/avocado/boot_linux_console.py
index ec07c64291..da9f3c4232 100644
--- a/tests/avocado/boot_linux_console.py
+++ b/tests/avocado/boot_linux_console.py
@@ -452,6 +452,7 @@ def test_arm_raspi2_uart0(self):
         """
         self.do_test_arm_raspi2(0)
 
+    @skip("pdel@fb.com: This test is flaky")
     def test_arm_raspi2_initrd(self):
         """
         :avocado: tags=arch:arm
-- 
2.38.2

