diff --git a/smbus.c b/smbus.c
index 431a6cb..77d858c 100644
--- a/smbus.c
+++ b/smbus.c
@@ -14,7 +14,6 @@
 
 #define pr_fmt(x) "smbus: " x
 
-#include <i2c/smbus.h>
 #include <linux/i2c-dev.h>
 #include <linux/i2c.h>
 #include <sys/ioctl.h>
diff --git a/tests/test_smbus.c b/tests/test_smbus.c
index e57b2b4..dd6e4f9 100644
--- a/tests/test_smbus.c
+++ b/tests/test_smbus.c
@@ -6,7 +6,6 @@
 #include "config.h"
 #endif
 
-#include <i2c/smbus.h>
 #include <linux/i2c-dev.h>
 #include <linux/i2c.h>
 #include "libmctp-alloc.h"
