From 08679d4c3cf00cd6d062f3c63d87232f31b9fb58 Mon Sep 17 00:00:00 2001
From: Howard Chiu <howard.chiu@quantatw.com>
Date: Fri, 18 Dec 2020 18:18:49 +0800
Subject: [PATCH] Increase the size of packet buffer

---
 smbus.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/smbus.c b/smbus.c
index 70bea4e..90ac6db 100644
--- a/smbus.c
+++ b/smbus.c
@@ -37,7 +37,7 @@
 #define SMBUS_COMMAND_CODE_SIZE 1
 #define SMBUS_LENGTH_FIELD_SIZE 1
 #define SMBUS_ADDR_OFFSET_SLAVE 0x1000
-#define SMBUS_MAX_PKT_PAYLOAD_SIZE 64
+#define SMBUS_MAX_PKT_PAYLOAD_SIZE 1480
 
 #ifdef I2C_M_HOLD
 static struct mctp_smbus_extra_params active_mux_info = { .fd = -1,
-- 
2.7.4

