From 9792590a8baf0d247e7503d358d67626e896f696 Mon Sep 17 00:00:00 2001
From: Tao Ren <taoren@meta.com>
Date: Thu, 29 May 2025 09:21:14 +0000
Subject: [PATCH] [PATCH] setup_i2c.service restart on failure

---
 setup_i2c.service | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/setup_i2c.service b/setup_i2c.service
index 699f1b32130..c29cdd87d48 100644
--- a/setup_i2c.service
+++ b/setup_i2c.service
@@ -29,6 +29,8 @@ DefaultDependencies=no
 Type=oneshot
 RemainAfterExit=yes
 ExecStart=/usr/local/bin/setup_i2c.sh
+Restart=on-failure
+RestartSec=3
 
 [Install]
 WantedBy=basic.target
-- 
2.34.1

