From b549960581862066c174e00cb0bdd704167eafc0 Mon Sep 17 00:00:00 2001
From: Christopher Covington <cov@fb.com>
Date: Mon, 17 Sep 2018 18:12:47 -0700
Subject: [PATCH] Allow link-local source-address over any interface

---
 addrmatch.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/addrmatch.c b/addrmatch.c
index 70b050e..a179159 100644
--- a/addrmatch.c
+++ b/addrmatch.c
@@ -445,6 +445,11 @@ addr_match_cidr_list(const char *addr, const char *_list)
 		debug2("%s: couldn't parse address %.100s", __func__, addr);
 		return 0;
 	}
+	/*
+	 * TODO Implement and deploy parsing of "fe80::%usb0/64" then remove this.
+	 */
+	try_addr.scope_id = 0;
+
 	if ((o = list = strdup(_list)) == NULL)
 		return -1;
 	while ((cp = strsep(&list, ",")) != NULL) {
-- 
2.17.1

