From c8bdfab7a3744b219c3b92e2551c16fcffbbf6f6 Mon Sep 17 00:00:00 2001
From: Peter Delevoryas <pdel@fb.com>
Date: Wed, 10 Aug 2022 14:43:25 -0700
Subject: [PATCH] Fix i2c-netdev2 send_async

Signed-off-by: Peter Delevoryas <pdel@fb.com>
---
 hw/misc/i2c-netdev2.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/hw/misc/i2c-netdev2.c b/hw/misc/i2c-netdev2.c
index f8c99277da..a630a28b0a 100644
--- a/hw/misc/i2c-netdev2.c
+++ b/hw/misc/i2c-netdev2.c
@@ -177,6 +177,9 @@ static int i2c_netdev2_handle_event(I2CSlave *i2c, enum i2c_event event)
         printf("%s: RECV UNIMPLEMENTED\n", __func__);
         abort();
     case I2C_START_SEND:
+        printf("%s: SEND UNIMPLEMENTED\n", __func__);
+        abort();
+    case I2C_START_SEND_ASYNC:
         // printf("%s: SENDING START MESSAGE PACKET\n", __func__);
         memset(start_msg, 0, sizeof(start_msg));
         start_msg[0] = tx_addr;
