diff -ur a/Source/TPMFactoryUpd/makefile b/Source/TPMFactoryUpd/makefile
--- a/Source/TPMFactoryUpd/makefile
+++ b/Source/TPMFactoryUpd/makefile
@@ -13,9 +13,9 @@
 #
 # The makefile uses the gcc compiler.
 #
-CC=$(CROSS_COMPILE)gcc
-STRIP=$(CROSS_COMPILE)strip
-AR=$(CROSS_COMPILE)ar
+CC ?= gcc
+STRIP ?= strip
+AR ?= ar
 #FPACK+= -fpack-struct # Don't pack structs globally. This would crash OpenSSL decrypt operation
 
 # Basic compiler options
