import http://sourceware.org/ml/libc-ports/2007-12/msg00000.html

Upstream-Status: Pending

Index: libc/bits/stdio-lock.h
===================================================================
--- libc.orig/bits/stdio-lock.h	2009-10-28 14:34:19.000000000 -0700
+++ libc/bits/stdio-lock.h	2009-10-28 14:34:54.000000000 -0700
@@ -50,6 +50,8 @@ __libc_lock_define_recursive (typedef, _
   _IO_cleanup_region_start ((void (*) (void *)) _IO_funlockfile, (_fp));      \
   _IO_flockfile (_fp)
 
+# define _IO_acquire_lock_clear_flags2(_fp) _IO_acquire_lock (_fp)
+
 # define _IO_release_lock(_fp) \
   _IO_funlockfile (_fp);						      \
   _IO_cleanup_region_end (0)
